parser/src/parser/mod.rs

```diff
@@ -1,7 +1,214 @@
-#![allow(clippy::all)]
-lalrpop_mod!(grammar, "/parser/grammar.rs");
+// Simple macro used in the grammar definition for constructing spans
+macro_rules! span {
+    ($l:expr, $r:expr) => {
+        miden_diagnostics::SourceSpan::new($l, $r)
+    };
+    ($i:expr) => {
+        miden_diagnostics::SourceSpan::new($i, $i)
+    };
+}
 
-pub use grammar::SourceParser;
+lalrpop_mod!(
+    #[allow(clippy::all)]
+    grammar,
+    "/parser/grammar.rs"
+);
+
+use std::sync::Arc;
+
+use miden_diagnostics::{
+    CodeMap, Diagnostic, DiagnosticsHandler, Label, SourceIndex, SourceSpan, ToDiagnostic,
+};
+use miden_parsing::{Scanner, Source};
+
+use crate::ast;
+use crate::lexer::{Lexed, Lexer, LexicalError, Token};
+
+pub type Parser = miden_parsing::Parser<()>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParseError {
```

Comment:
Errors should be defined in the modules they are generated in, e.g. `ParseError` should be defined near the parser.

Reply:
Ignore my previous comment, I misunderstood what you were suggesting. I would say that in general it depends on how crowded the root module is. In this case, there is nothing else defined in the root except the `ParseError` type, so having a nested module doesn't really have any benefits and adds yet another file for `rustc` to read from disk, parse, and compile. I tend to nest errors in their own module only if they have unusually verbose implementations, or the root module to which they belong is already quite noisy/large.

Reply:
Primary motivation for my comment was that most of `mod.rs` file seems to be taken up by `ParserError` and related code, and it felt like it would be cleaner to move it into its own module (even if this means `mod.rs` would have very little of actual code left). But this is also not a strong preference - so, I'm fine keeping it here.